examples/Fine_tuning_for_function_calling.ipynb

```diff
@@ -0,0 +1,1248 @@
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Play pre-recorded audio message \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"reject_request\",\n",
+      "  \"arguments\": \"{}\"\n",
+      "}\n",
+      "Initiate live-streaming on social media \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"control_camera\",\n",
+      "  \"arguments\": \"{\\n  \\\"mode\\\": \\\"video\\\",\\n  \\\"duration\\\": 0\\n}\"\n",
+      "}\n",
+      "Scan environment for heat signatures \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"control_camera\",\n",
+      "  \"arguments\": \"{\\n  \\\"mode\\\": \\\"photo\\\"\\n}\"\n",
+      "}\n",
+      "Enable stealth mode \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"set_drone_lighting\",\n",
+      "  \"arguments\": \"{\\n  \\\"mode\\\": \\\"off\\\"\\n}\"\n",
+      "}\n",
+      "Change drone's paint job color \n",
+      "\n",
+      "{\n",
+      "  \"name\": \"configure_led_display\",\n",
+      "  \"arguments\": \"{\\n  \\\"pattern\\\": \\\"solid\\\",\\n  \\\"color\\\": \\\"blue\\\"\\n}\"\n",
+      "}\n"
+     ]
+    }
+   ],
+   "source": [
+    "for prompt in challenging_prompts:\n",
+    "  messages = []\n",
+    "  messages.append({\"role\": \"system\", \"content\": DRONE_SYSTEM_PROMPT})\n",
+    "  messages.append({\"role\": \"user\", \"content\": prompt})\n",
+    "  completion = get_function_call(model=\"gpt-3.5-turbo\",messages=messages,functions=function_list)\n",
+    "  print(prompt,'\\n')\n",
+    "  try:\n",
+    "    print(completion.function_call)\n",
+    "  except:\n",
+    "    print(completion.content)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now we run into some problems.\n",
+    "The model here should reject all of these requests, as they are impossible given the functions, however instead the model calls functions that are somewhat related to the request, but incorrect. The model sets the camera to video when asked to begin 'live streaming to social media', and changes the LED's to blue when asked to 'change the paint color'... \\\n",
+    "Even with prompt engineering (in the prompt we say to decline requests when there is any ambiguity or doubt), the function calling is not working well. So this is a good opportuntiy for fine tuning for function calling."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Generating synthetic data"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Helper functions"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We want to generate every invocation of every function, so that we have\n",
+    "full coverage of all potential invocations to create synthetic data for. Then, we will use GPT-4 to come up with prompts that would call each invocation, and we will use that prompt - function invocation pair as training data."
```

Comment:
Might want to motivate this – e.g. To perform effective fine-tuning we need correctly labeled data. We could manually come up with examples and label the data... or we could generate synthetic data with the help of GPT4. One way to generate a comprehensive dataset it to...